### ISSUE TITLE
CancellationTokenCallback never fired when SSE connection ends in Azure Web App

### ISSUE THREAD

**Issue by henry-chris** [tpeczek/Lib.AspNetCore.ServerSentEvents#35]
**Configuration:**
Linux Web App
DOTNETCORE|3.1    (For LinuxFXVersion)  I can't tell specifically what runtime it is yet.

Build uses Core 3.1.2 to build project.
Use 'event-source-polyfill'  npm pacakge in browser.  (Both 1.0.11 and 1.0.12)

Library version is 4.0.0, but I've also tested 4.1.0.

**Description:**
In the SSE Middleware the pipeline looks like this -
![image](https://user-images.githubusercontent.com/4165721/79612079-92504900-80c1-11ea-9529-80dc53a51320.png)

Within `WaitAsync` is where you seem to wait for a cancellation of the HTTP request.

![image](https://user-images.githubusercontent.com/4165721/79612131-b0b64480-80c1-11ea-890c-3e234d450830.png)

When deployed to Azure, the `CancellationTokenCallback` is never invoked for any connection.  

When run locally (Windows Machine with Kestrel) everything works and the callback is hit correctly.

**Notes:**
Have been using the library for many months on a Linux App Service in very close to the same way.

Currently, Production deployment works perfectly.  This was deployed on 3/26 and runs on a Linux Web App with Core 3.1.2.  I can't tell if the App Service Runtime is different than the current runtime, though.

At some point between 3/26 and 4/08 (That's when my logs stop) something seems to have change.  When I take the exact artifact which is deployed in Production and then re-deploy it in another app service I get the same result as any new builds.  Clients are not being Cancelled.

**Thoughts:**
The only difference I can find so far is that in the production app, there is no 'server' header on the response.  

In any new deployment the header  `server: Kestrel` exists even though I explicitly instruct Kestrel not to add it. This leads me to believe something has changed that I can't find.  No logs or errors present.

![image](https://user-images.githubusercontent.com/4165721/79612626-a6487a80-80c2-11ea-8745-5b90347d7142.png)

Other possibilities include different runtime or some type of update to Azure infrastructure.

Or something I'm missing in code, but this seems less likely right now.

I have pulled down your code and am just putting logging statements everywhere to ensure that it is failing in the place where I expect when deployed.  Then building locally and using that to run in an app service.

Next step is to convert to a Windows App Service to see if that changes the outcome.

It doesn't **seem** to be your library causing the issue, but I could be wrong.  At the time I originally discovered this, I thought it more likely.

**Comment by henry-chris** [comment]
Not sure how much you can help or what your plan was.  Just giving some updates.  

I need to redeploy in Linux with some logging modifications done to your code so I can verify exactly the code which is failing (in case something weird is happening).

For now, it seems like a Linux App Service Specific problem, although I haven't been able to rule some things out.  I remember this being an issue before .Net Core 2.2...but it has been working for a long time.  May need to open a ticket with Azure or Asp.net Core somewhere.

If you have any thoughts, let me know.

**Test 1: In-Process IIS on Windows App Service** 
Confirmed with 'server: Microsoft IIS/10.0'   (something close to that)
Disconnect Events **do** trigger.

**Test 2: Out of Process IIS on Windows App Service** 
Confirmed with 'server: Kestrel' header
Disconnect Events **do** trigger.

**Test 3: Self-Contained vs. Runtime Dependent on Windows** 
Disconnect Events **do** trigger for both.

**Test 4: Self-Contained vs. Runtime Dependent on Linux** 
Disconnect Events **do not** trigger for either.

**Comment by henry-chris** [comment]
Confirmed it has nothing to do with the library as I reproduced it in a minimal MVC app today with HttpContext.RequestAborted.ThrowIfCancellationRequested().

Works in windows, not in Linux.

Opened a ticket with Azure.  Let me know if you want me to close this or do anything else.

**Comment by tpeczek** [comment]
Yes, I don't think it's something which can be fixed by library.

But if you want we can try discuss some application level workaround which can help mitigate that in your context.

**Comment by henry-chris** [comment]
I appreciate it it.  

I'm hoping Azure has an answer soon.  It feels like something strange is happening since it's been working fine for about a year and suddenly stopped out of nowhere.

**My workaround around options are as follows (that I know of):**
1. Switch to Windows App Service
2. Implement some type of polling from the client
3. Implement some type of session timeout/invalidation
4. Attempt to use WebSockets since it's two-way communication

If you have any other suggestions let me know.  Otherwise, feel free to close this.

**Comment by tpeczek** [comment]
I'm thinking about using keepalive to workaround this.

I didn't had chance to reproduce the issue my self, but if keepalives are failing with a specific exception I should be able to catch that in `ServerSentEventsKeepaliveService` and disconnect client there. I would have to know if and how keepalives fail in this scenario, if you still have a repro environment can you set keepalives to always and see what is being thrown?

**Comment by henry-chris** [comment]
There is no exception that I can find.  Or it's being swallowed somewhere.

Sadly my minimal repro environment was just to use an Asp.net Core template and create a controller that delayed response for 30 seconds.  Then I make a call to the controller from the browser and close the browser window at around 5-10 seconds.  (All in Linux App Service)

I tried multiple ways to listen for the cancellation token / RequestAborted, but it never triggers.

I will inject some logging and add some more exception handling with the KeepAlive service and see what I can find.  I honestly have not tried that yet.

I will likely wait a few more days as MS is still conferring over the help ticket.

**Comment by henry-chris** [comment]
Currently waiting on MS response.

The problem that I have now is that there is no way to manually remove a client.  No matter what workaround I use the clients will continue to build up in the dictionary as there is no public facing way to remove clients.

I guess I might be able to use reflection to trigger `ServerSentEventsService.RemoveClient` and then trigger the Disconnect event manually, but that's very sub-optimal if it works.

I'm not seeing any type of error at all when the `KeepAliveService `continues to push messages through even very old clients.

**Comment by tpeczek** [comment]
The problem is much deeper than just removing clients from the dictionary, there is a bunch of (dormant but still eating resources) threads hanging there.

If you have a logical conditions based on which you can determine which clients should be disconnected, they you may try aborting th connection itself. You can grab the `HttpContext` for a client in `OnClientConnected` event:

```cs
options.OnClientConnected = (service, args) =>
{
    _clientsContexts.Add(args.Client.Id, args.Request.HttpContext);
};
```

When the logical conditions are met, you can try to call `Abort()`.

```cs
_clientsContexts[args.Client.Id].Abort();
```

Depending on the nature of the bug in ASP.NET Core, it may do the trick.

This is something I would never suggest in general, because it's basically disconnecting client form server side which is not supported in SSE (in case of a connected browser, this will most likely trigger automatic reconnect). In your case this might be helpful.

**Comment by henry-chris** [comment]
Crazy.  I think I've found and fixed the problem.

If you create a Linux Web App and add _any_ CORS urls/domains then RequestAborted stops working.  Maybe the connection happens at a different point when you use the Web App CORS settings?

Anyways, I had recently added CORS url for my CDN and removing that fixed it.  Never would have guessed.

Thanks for the help and I'm closing this now.

![image](https://user-images.githubusercontent.com/4165721/80973935-a4322b80-8de5-11ea-83b3-827792727c34.png)
